src/backend/distributed/connection/connection_management.c

```diff
@@ -437,184 +457,378 @@ ShutdownConnection(MultiConnection *connection)
+				(MultiConnectionState *) event->user_data;
+
+			if (event->events & WL_POSTMASTER_DEATH)
 			{
-				FD_SET(selectFileDescriptor, &writeFileDescriptorSet);
+				ereport(ERROR, (errmsg("postmaster was shut down, exiting")));
 			}
 
-			pollResult = (select) (selectFileDescriptor + 1, &readFileDescriptorSet,
-								   &writeFileDescriptorSet, &exceptionFileDescriptorSet,
-								   &selectTimeout);
-#endif /* HAVE_POLL */
-
-			if (pollResult == 0)
+			if (event->events & WL_LATCH_SET)
 			{
-				/*
-				 * Timeout exceeded. Two things to do:
-				 * - check whether any interrupts arrived and handle them
-				 * - check whether establishment for connection already has
-				 *   lasted for too long, stop waiting if so.
-				 */
+				ResetLatch(MyLatch);
+
 				CHECK_FOR_INTERRUPTS();
 
-				if (TimestampDifferenceExceeds(connection->connectionStart,
-											   GetCurrentTimestamp(),
-											   NodeConnectionTimeout))
+				if (InterruptHoldoffCount > 0 && (QueryCancelPending || ProcDiePending))
 				{
-					ereport(WARNING, (errmsg("could not establish connection after %u ms",
-											 NodeConnectionTimeout)));
-
-					/* close connection, otherwise we take up resource on the other side */
-					PQfinish(connection->pgConn);
-					connection->pgConn = NULL;
-					break;
+					/*
+					 * because we can't break from 2 loops easily we need to not forget to
+					 * reset the memory context
+					 */
+					MemoryContextDelete(MemoryContextSwitchTo(oldContext));
+					return;
 				}
+
+				continue;
 			}
-			else if (pollResult > 0)
-			{
-				/*
-				 * IO possible, continue connection establishment. We could
-				 * check for timeouts here as well, but if there's progress
-				 * there seems little point.
-				 */
-				break;
-			}
-			else
+
+			switch (MultiConnectionStatePoll(connectionState))
 			{
-				int errorCode = errno;
-#ifdef WIN32
-				errorCode = WSAGetLastError();
-#endif
-				if (errorCode == EINTR)
+				case Rebuild:
 				{
-					/* Retrying, signal interrupted. So check. */
-					CHECK_FOR_INTERRUPTS();
+					waitEventSetRebuild = true;
+					break;
 				}
-				else
+
+				case Update:
 				{
-					/*
-					 * We ERROR here, instead of just returning a failed
-					 * connection, because this shouldn't happen, and indicates a
-					 * programming error somewhere, not a network etc. issue.
-					 */
-					ereport(ERROR, (errcode_for_socket_access(),
-									errmsg("poll()/select() failed: %m")));
+					uint32 eventMask = MultiConnectionStateEventMask(connectionState);
+					ModifyWaitEvent(waitEventSet, event->pos, eventMask, NULL);
+					break;
+				}
+
+				case None:
+				{
+					break;
 				}
 			}
 		}
+
+		if (eventCount == 0)
+		{
+			/*
+			 * timeout has occured, lets check there is an actual timeout, report if that
+			 * is the case and close all non-connected sockets
+			 */
+
+			bool timeoutOccured = CheckMultiConnectionStateTimeouts(connectionStates);
+			if (timeoutOccured)
```

Comment:
Checking all the different timers seems a bit weird. What if we just have one timer for this function?

Theoretically the connections could start at different times, but practically...